File: android/app/src/main/java/app/releaf/mobile/ui/components/PageHeaderControls.kt
/*
 * PageHeaderControls.kt
 *
 * Three small components that compose the top zone of PageDetailScreen:
 *
 *   - LeafEyebrow         — leaf glyph + tracked uppercase label
 *   - PageViewToggle      — two-tab segmented pill (list / grid)
 *   - PageOverflowButton  — round kebab button
 *
 * Decorative for now — the toggle is wired to a state holder so the
 * parent tracks the view mode, but routing list-vs-grid to actual
 * layouts is a follow-up. Same for the overflow: it accepts an action
 * lambda but the action sheet hasn't been designed yet.
 *
 * Mirrors the iOS `PageHeaderControls.swift`.
 */

package app.releaf.mobile.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.GridView
import androidx.compose.material.icons.filled.MoreVert
import androidx.compose.material.icons.automirrored.outlined.FormatListBulleted
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import app.releaf.mobile.ui.theme.AppAccent
import app.releaf.mobile.ui.theme.AppColors
import app.releaf.mobile.ui.theme.AppRadius
import app.releaf.mobile.ui.theme.AppSpacing
import app.releaf.mobile.ui.theme.AppTypography

// ---------- LeafEyebrow ----------

@Composable
fun LeafEyebrow(
    label: String,
    modifier: Modifier = Modifier,
    glyphTint: androidx.compose.ui.graphics.Color? = null,
    labelTint: androidx.compose.ui.graphics.Color? = null,
) {
    Row(
        modifier              = modifier,
        verticalAlignment     = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(AppSpacing.s1),
    ) {
        LeafDropletGlyph(
            tint = glyphTint ?: AppColors.ThemeGreenPrimary,
            size = 11.dp,
        )
        Text(
            text     = label.uppercase(),
            style    = AppTypography.Eyebrow,
            color    = labelTint ?: AppColors.ThemeGreenDeep,
            maxLines = 1,
            overflow = TextOverflow.Ellipsis,
        )
    }
}

// ---------- PageViewMode ----------

enum class PageViewMode(
    val displayName: String,
    val icon: ImageVector,
) {
    List("List", Icons.AutoMirrored.Outlined.FormatListBulleted),
    Grid("Grid", Icons.Filled.GridView),
}

// ---------- PageViewToggle ----------

@Composable
fun PageViewToggle(
    selected: PageViewMode,
    onSelect: (PageViewMode) -> Unit,
    modifier: Modifier = Modifier,
) {
    Row(
        modifier = modifier
            .clip(RoundedCornerShape(AppRadius.pill))
            .background(AppColors.CardSolid.copy(alpha = 0.6f))
            .border(1.dp, AppColors.BorderDefault, RoundedCornerShape(AppRadius.pill))
            .padding(3.dp),
        horizontalArrangement = Arrangement.spacedBy(2.dp),
        verticalAlignment     = Alignment.CenterVertically,
    ) {
        PageViewMode.values().forEach { mode ->
            ViewToggleSegment(
                mode       = mode,
                isSelected = selected == mode,
                onClick    = { onSelect(mode) },
            )
        }
    }
}

@Composable
private fun ViewToggleSegment(
    mode: PageViewMode,
    isSelected: Boolean,
    onClick: () -> Unit,
) {
    val tint = if (isSelected) AppColors.OnAccent else AppColors.TextSecondary
    val bg   = if (isSelected) AppAccent.primary else androidx.compose.ui.graphics.Color.Transparent

    Box(
        modifier = Modifier
            .clip(CircleShape)
            .background(bg)
            .clickable(onClick = onClick)
            .width(28.dp)
            .height(22.dp),
        contentAlignment = Alignment.Center,
    ) {
        Icon(
            imageVector        = mode.icon,
            contentDescription = mode.displayName,
            tint               = tint,
            modifier           = Modifier.size(13.dp),
        )
    }
}

// ---------- PageOverflowButton ----------

/**
 * Round overflow button that opens a Material3 [DropdownMenu] of the
 * parent's items. Caller provides the items as a `ColumnScope`
 * content lambda — typically a stack of [androidx.compose.material3.DropdownMenuItem]
 * rows. Anchor + expanded state are managed inside the component so
 * the call site only ever has to think about the actions.
 */
@Composable
fun PageOverflowButton(
    modifier: Modifier = Modifier,
    content: @Composable ColumnScope.() -> Unit,
) {
    var expanded by remember { mutableStateOf(false) }

    Box(modifier = modifier) {
        Box(
            modifier = Modifier
                .clip(CircleShape)
                .background(AppColors.CardSolid.copy(alpha = 0.6f))
                .border(1.dp, AppColors.BorderDefault, CircleShape)
                .clickable { expanded = true }
                .size(28.dp),
            contentAlignment = Alignment.Center,
        ) {
            Icon(
                imageVector        = Icons.Filled.MoreVert,
                contentDescription = "More",
                tint               = AppColors.TextPrimary,
                modifier           = Modifier.size(15.dp),
            )
        }
        DropdownMenu(
            expanded         = expanded,
            onDismissRequest = { expanded = false },
            content          = content,
        )
    }
}

// ---------- Preview ----------

@Preview(showBackground = true, backgroundColor = 0xFFF5EEDF, widthDp = 360)
@Composable
private fun PageHeaderControlsPreview() {
    val mode = remember { mutableStateOf(PageViewMode.Grid) }
    Row(
        modifier              = Modifier.padding(AppSpacing.s4),
        verticalAlignment     = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(AppSpacing.s2),
    ) {
        LeafEyebrow("notepad · today", modifier = Modifier.padding(end = AppSpacing.s4))
        PageViewToggle(selected = mode.value, onSelect = { mode.value = it })
        PageOverflowButton {
            androidx.compose.material3.DropdownMenuItem(
                text = { Text("Move to notebook") },
                onClick = {},
            )
            androidx.compose.material3.DropdownMenuItem(
                text = { Text("Apply template") },
                onClick = {},
            )
            androidx.compose.material3.DropdownMenuItem(
                text = { Text("Archive page") },
                onClick = {},
            )
        }
    }
}

File: shared/ios/ReleafCore/Sources/ReleafCoreDesignSystem/Components/PageHeaderControls.swift
/*
 * PageHeaderControls.swift
 *
 * Three small components that compose the top zone of PageDetailView:
 *
 *   - LeafEyebrow         — leaf glyph + tracked uppercase label
 *   - PageViewToggle      — two-tab segmented pill (list / grid)
 *   - PageOverflowButton  — round kebab button
 *
 * Decorative for now — the toggle is wired to a Binding so the parent
 * tracks the view mode, but routing list-vs-grid to actual layouts is
 * a follow-up. Same for the overflow: it accepts an action closure
 * but the action sheet hasn't been designed yet.
 *
 * Mirrors the Android `LeafEyebrow.kt` / `PageViewToggle.kt` /
 * `PageOverflowButton.kt`.
 */

import SwiftUI

// MARK: - LeafEyebrow

public struct LeafEyebrow: View {
    public let label: String
    public let onTap: (() -> Void)?
    /// Optional tint pair — when supplied, overrides the default
    /// green leaf color for the glyph and the eyebrow label. Used
    /// to thread per-notebook / per-shelf color into the header so
    /// the eyebrow matches the row chip the user just tapped.
    public let glyphTint: Color?
    public let labelTint: Color?

    public init(
        _ label: String,
        glyphTint: Color? = nil,
        labelTint: Color? = nil,
        onTap: (() -> Void)? = nil
    ) {
        self.label = label
        self.glyphTint = glyphTint
        self.labelTint = labelTint
        self.onTap = onTap
    }

    public var body: some View {
        let glyphAndText = HStack(spacing: AppSpacing.s1) {
            LeafDropletGlyph(
                tint: glyphTint ?? AppColors.themeGreenPrimary,
                size: 11
            )
            Text(label.uppercased())
                .font(AppText.eyebrow)
                .tracking(AppLetterSpacing.eyebrow)
                .foregroundStyle(labelTint ?? AppColors.themeGreenDeep)
        }
        .accessibilityElement(children: .combine)
        .accessibilityLabel(Text(label))

        if let onTap {
            Button(action: onTap) { glyphAndText }
                .buttonStyle(.plain)
                .accessibilityHint("Go back")
        } else {
            glyphAndText
        }
    }
}

// MARK: - PageViewMode

public enum PageViewMode: String, CaseIterable, Codable, Sendable {
    case list
    case grid

    public var systemIcon: String {
        switch self {
        case .list: return "list.bullet"
        case .grid: return "square.grid.2x2"
        }
    }
}

// MARK: - PageViewToggle

public struct PageViewToggle: View {
    @Binding private var selected: PageViewMode

    @Environment(\.accentPalette) private var accent

    public init(selected: Binding<PageViewMode>) {
        self._selected = selected
    }

    public var body: some View {
        HStack(spacing: 2) {
            ForEach(PageViewMode.allCases, id: \.self) { mode in
                Button {
                    withAnimation(.spring(response: 0.25, dampingFraction: 0.85)) {
                        selected = mode
                    }
                } label: {
                    Image(systemName: mode.systemIcon)
                        .font(.system(size: 11, weight: selected == mode ? .semibold : .regular))
                        .foregroundStyle(selected == mode
                                         ? AppColors.textOnAccent
                                         : AppColors.textSecondary)
                        .frame(width: 26, height: 22)
                        .background(
                            Capsule()
                                .fill(selected == mode ? accent.primary : Color.clear)
                        )
                        .contentShape(Rectangle())
                }
                .buttonStyle(.plain)
                .accessibilityLabel(Text(mode.rawValue.capitalized))
                .accessibilityAddTraits(selected == mode ? [.isSelected] : [])
            }
        }
        .padding(3)
        .background(AppColors.cardSolid.opacity(0.6))
        .overlay(
            Capsule().stroke(AppColors.borderDefault, lineWidth: 1)
        )
        .clipShape(Capsule())
    }
}

// MARK: - PageOverflowButton

/// Round overflow button that opens a SwiftUI Menu of the parent's
/// items. Caller provides the menu items as a `@ViewBuilder` closure
/// — typically a stack of `Button("Label") { action }` rows. The
/// button's chrome (size, fill, border, glyph) lives here so the
/// menu's call site only ever has to think about the actions.
public struct PageOverflowButton<MenuContent: View>: View {
    @ViewBuilder private let content: () -> MenuContent

    public init(@ViewBuilder content: @escaping () -> MenuContent) {
        self.content = content
    }

    public var body: some View {
        Menu {
            content()
        } label: {
            Image(systemName: "ellipsis")
                .rotationEffect(.degrees(90))
                .font(.system(size: 13))
                .foregroundStyle(AppColors.textPrimary)
                .frame(width: 28, height: 28)
                .background(
                    Circle().fill(AppColors.cardSolid.opacity(0.6))
                )
                .overlay(
                    Circle().stroke(AppColors.borderDefault, lineWidth: 1)
                )
        }
        .menuStyle(.borderlessButton)
        .accessibilityLabel(Text("More"))
    }
}

#if DEBUG
struct PageHeaderControls_Previews: PreviewProvider {
    struct Host: View {
        @State private var mode: PageViewMode = .grid
        var body: some View {
            VStack(alignment: .leading, spacing: AppSpacing.s4) {
                LeafEyebrow("notepad · today")
                HStack(spacing: AppSpacing.s2) {
                    PageViewToggle(selected: $mode)
                    PageOverflowButton {
                        Button("Move to notebook") {}
                        Button("Apply template") {}
                        Button("Duplicate") {}
                        Button("Archive page", role: .destructive) {}
                    }
                }
            }
            .padding()
            .background(AppColors.canvas)
            .accentPalette(AccentPalettes.green)
        }
    }
    static var previews: some View {
        Host().frame(width: 360)
    }
}
#endif
